### PR TITLE
APS-2756 - Manage PodDisruptionBudget for health-api on switchover

### PR DESCRIPTION
**Pull request by rustyjux**
# Description

From https://dpdd.atlassian.net/browse/APS-2756

Switchover was kicked off in prod on  when a node holding both Gold health-api pods was drained as part of Platform Services maintenance.

To avoid this situation in future, antiAffinity for these pods and a PodDisruptionBudget (minAvailable: 1) was added in [Merge pull request #221 from bcgov-dss/dev · bcgov-dss/api-serv-infra@e1a539f (github.com)](https://github.com/bcgov-dss/api-serv-infra/commit/e1a539f713946e2c709fda20b8f90c5d28968032#diff-eef93ca73e77c6c6deda2bf5d65d3c2b01f099bcdaf758b6e7cb210c2a37f630).

This change needs more work, as all PDBs are being left in place on both Gold and GoldDR and then the inactive cluster is sending alerts that the PDB is not satisfied.

AC:

- scale minAvailable along with the  health-api deployment (so that min 1 is available on active cluster, and 0 on passive)

---
Also fixes:
- Update switchover-agent to check status of pipeline when considering it being done (https://dpdd.atlassian.net/browse/APS-3006) in https://github.com/bcgov/switchover-agent/pull/18/commits/9c3d9e04e9f5d1dc457232e6bb45b1f9447941ba

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation (non-breaking change with enhancements to documentation)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

I'm not sure we have any tests for switchover agent? I have tested by running a switchover cycle to GoldDR and back to Gold on dev.

I have not tested locally.

- [ ] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

This change includes an upgrade to the kubernetes python client version (to allow use of PolicyV1Api, introduced in k8s v1.21 - OCP Gold is currently running v1.27.16). I upgraded to latest `kubernetes = "^31.0.0"`

### CODE REVIEW

**Comment by rustyjux** [comment]
To confirm - the health-api is meant to always be up in GoldDR (we don't scale it down in case of failover)? Only health-api in Gold gets scaled down to 0?

File: src/clients/kube.py [hunk 1]
```diff
@@ -264,3 +270,25 @@ def init_apps_client(py_env: str):
...
+    try:
+        body = {
+            "spec": {
+                "minAvailable": min_available
+            }
+        }
+        v1.patch_namespaced_pod_disruption_budget(name, namespace, body)
+        logger.debug(f"PDB {name} updated successfully")
+    except ApiException as e:
+        logger.error(f"Exception when updating PDB: {e}")
```
Comment (ikethecoder) [inline src/clients/kube.py:289]: I would add "raise" in there too after the logger statement, so as not to hide the exception.

File: src/clients/kube.py [hunk 2]
```diff
@@ -133,6 +134,11 @@ def scale(namespace: str, kind: str, name: str, replicas: int, py_env: str):
             v1.patch_namespaced_stateful_set_scale(
                 name, namespace, body)
 
+        if name == switchover_config.get('deployment_health_api'):
```
Comment (ikethecoder) [inline src/clients/kube.py]: This logic should really be taken out and just call update_pdb in places like https://github.com/bcgov/switchover-agent/blob/main/src/transitions/initiate_primary.py#L29
Reason is that kube.py provides convenience functions for using the kube api and this logic is too specific to the switchover-agent.

**Comment by rustyjux** [comment]
Did some reading of https://github.com/bcgov-dss/api-serv-infra/blob/docs/ops/docs/playbooks/SWITCHOVER-AGENT.md#steps-performed-by-switchover-agent and it looks like `active-passive` --> `golddr-maintenance` is the only transition that would result in a scale down of health in GoldDR. 

`golddr-maintenance` is the odd ball we don't regularly use but could occasionally go to manually if we knew GoldDR was down:
> golddr-maintenance : This scenario is that Gold is Primary and Gold DR is detached for maintenance/testing; effectively independent. Could be called active-only.

I tried to test this by transitioning --> `golddr-maintenance` on GoldDR (does nothing if done on Gold - https://vscode.dev/github/bcgov/switchover-agent/blob/dev/src/logic.py#L372), but it failed, so this change remains untested for the case of scaling health-api on GoldDR:

```
2024-10-04 16:02:53,608 [DEBUG] clients.kube Event: MODIFIED ConfigMap switchover-state-dev
2024-10-04 16:02:53,609 [INFO ] clients.kube {
"last_stable_state": "active-passive",
"last_stable_state_ts": "2024-10-03T17:20:36.889531",
"transition": "golddr-maintenance"
}
2024-10-04 16:02:53,609 [INFO ] logic logic switchover_state
2024-10-04 16:02:53,609 [INFO ] logic transitioning to golddr-maintenance
2024-10-04 16:02:53,609 [ERROR] logic Unknown error in logic. initiate_passive_maintenance() takes 3 positional arguments but 4 were given
2024-10-04 16:05:07,824 [DEBUG] clients.kube Event: MODIFIED ConfigMap switchover-state-dev
2024-10-04 16:05:07,825 [INFO ] clients.kube {
"last_stable_state": "active-passive",
"last_stable_state_ts": "2024-10-03T17:20:36.889531",
"transition": ""
}
Traceback (most recent call last):
File "/app/src/logic.py", line 377, in handler
initiate_passive_maintenance(self,
TypeError: initiate_passive_maintenance() takes 3 positional arguments but 4 were given
```

@ikethecoder should we open a bug ticket to investigate `golddr-maintenance`?

**Comment by ikethecoder** [comment]
Yeah, we should get `golddr-maintenance` fixed.

**Review by ikethecoder — CHANGES_REQUESTED** [review]

File: src/logic.py [hunk 3]
```diff
@@ -78,14 +78,17 @@ def handler(self, cluster: str, namespace: str, label_selector: str, patroni_loc
...
+                                            % status_reason)
                                 logger.info("Tekton Start: %s" %
                                             self.pipeline['start_ts'])
                                 logger.info("Tekton   End: %s" %
                                             datetime.datetime.now())
                                 # set the maintenance mode appropriately
                                 if self.pipeline['maintenance']:
                                     maintenance_on()
+                                elif status_reason == "Failed" or status_reason == "PipelineRunCancelled":
+                                    logger.info("Pipeline Failed or Cancelled - keeping maintenance on")
```
Comment (ikethecoder) [inline src/logic.py]: logger.error

**Review by ikethecoder — APPROVED** [review]

File: src/transitions/initiate_standby.py [hunk 4]
```diff
@@ -1,27 +1,27 @@
...
                             py_env, 'gold-standby')
 
 
 def initiate_passive_standby(logic_context, py_env: str):
     set_in_recovery(False, py_env)
-    scale(config.get('kube_health_namespace'), 'deployment',
-          config.get('deployment_health_api'), 2, py_env)
+    scale_health_api(config.get('kube_health_namespace'), 
+                     config.get('deployment_health_api'), 
+                     s, py_env)
```
Comment (ikethecoder) [inline src/transitions/initiate_standby.py]: I think there is an `s` that should be a 2

**Review by ikethecoder — CHANGES_REQUESTED** [review]

File: src/transitions/initiate_standby.py (hunk 4 above)
Comment (rustyjux) [reply src/transitions/initiate_standby.py]: good catch. use suggestions!
```suggestion
                     2, py_env)
```

**Comment by rustyjux** [comment]
Observed pdb successfully changing from minAvailable 1 --> 0 --> 1 in Gold during failover cycle. Also observed maintenance staying on when a pipeline failed.

**Pull request merged**